2.2 regression: can not bind member function with lifted access

Dear Pybind11 developers, here is another 2.2 regression issue (exact same code was working at a22dd2d 

Right now it seems not possible to bind member function which access was lifter to `public` with `using` declaration. Please see minimal example and error message below.

Thanks,

```C++
#include <pybind11/pybind11.h>

struct Base
{
	void foo() {}
};

class Child : private Base
{
public:
	using Base::foo;
};


PYBIND11_PLUGIN(example) {
    pybind11::module m("example", "pybind11 example plugin");

	pybind11::class_<Child>(m, "Child")
        .def("foo", &Child::foo);

    return m.ptr();

	Child c; c.foo();
}
```

Error:
```
/home/benchmark/src/pybind11/include/pybind11/pybind11.h:1000:93: error: cannot cast private base class 'Base' to 'Child'
auto method_adaptor(Return (Class::*pmf)(Args...)) -> Return (Derived::*)(Args...) { return pmf; }
                                                                                            ^
/home/benchmark/src/pybind11/include/pybind11/pybind11.h:1078:25: note: in instantiation of function template specialization 'pybind11::method_adaptor<Child, void, Base>' requested here
        cpp_function cf(method_adaptor<type>(std::forward<Func>(f)), name(name_), is_method(*this),
                        ^
access-lifting.cpp:20:10: note: in instantiation of function template specialization 'pybind11::class_<Child>::def<void (Base::*)()>' requested here
        .def("foo", &Child::foo);
         ^
access-lifting.cpp:8:15: note: declared private here
class Child : private Base
              ^~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/include/pybind11/detail/common.h
+++ b/include/pybind11/detail/common.h
@@ -581,6 +581,11 @@ template <typename T, typename... Us> using deferred_t = typename deferred_type<
 template <typename Base, typename Derived> using is_strict_base_of = bool_constant<
     std::is_base_of<Base, Derived>::value && !std::is_same<Base, Derived>::value>;
 
+/// Like is_base_of, but also requires that the base type is accessible (i.e. that a Derived pointer
+/// can be converted to a Base pointer)
+template <typename Base, typename Derived> using is_accessible_base_of = bool_constant<
+    std::is_base_of<Base, Derived>::value && std::is_convertible<Derived *, Base *>::value>;
+
 template <template<typename...> class Base>
 struct is_template_base_of_impl {
     template <typename... Us> static std::true_type check(Base<Us...> *);

--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -997,10 +997,18 @@ template <typename /*Derived*/, typename F>
 auto method_adaptor(F &&f) -> decltype(std::forward<F>(f)) { return std::forward<F>(f); }
 
 template <typename Derived, typename Return, typename Class, typename... Args>
-auto method_adaptor(Return (Class::*pmf)(Args...)) -> Return (Derived::*)(Args...) { return pmf; }
+auto method_adaptor(Return (Class::*pmf)(Args...)) -> Return (Derived::*)(Args...) {
+    static_assert(detail::is_accessible_base_of<Class, Derived>::value,
+        "Cannot bind an inaccessible base class method; use a lambda definition instead");
+    return pmf;
+}
 
 template <typename Derived, typename Return, typename Class, typename... Args>
-auto method_adaptor(Return (Class::*pmf)(Args...) const) -> Return (Derived::*)(Args...) const { return pmf; }
+auto method_adaptor(Return (Class::*pmf)(Args...) const) -> Return (Derived::*)(Args...) const {
+    static_assert(detail::is_accessible_base_of<Class, Derived>::value,
+        "Cannot bind an inaccessible base class method; use a lambda definition instead");
+    return pmf;
+}
 
 template <typename type_, typename... options>
 class class_ : public detail::generic_type {

